File: backend/features/offers/offers-core/src/main/java/de/bennyboer/kicherkrabbe/offers/PriceHistoryEntry.java
package de.bennyboer.kicherkrabbe.offers;

import de.bennyboer.kicherkrabbe.money.Money;
import lombok.AllArgsConstructor;
import lombok.Value;

import java.time.Instant;

import static de.bennyboer.kicherkrabbe.commons.Preconditions.notNull;
import static lombok.AccessLevel.PRIVATE;

@Value
@AllArgsConstructor(access = PRIVATE)
public class PriceHistoryEntry {

    Money price;

    Instant timestamp;

    public static PriceHistoryEntry of(Money price, Instant timestamp) {
        notNull(price, "Price must be given");
        notNull(timestamp, "Timestamp must be given");

        return new PriceHistoryEntry(price, timestamp);
    }

}

File: frontend/customer/src/app/offers/model/price-history-entry.ts
import { validateProps } from "@kicherkrabbe/shared";
import { Money } from "./money";

export class PriceHistoryEntry {
	readonly price: Money;
	readonly timestamp: Date;

	private constructor(props: { price: Money; timestamp: Date }) {
		validateProps(props);

		this.price = props.price;
		this.timestamp = props.timestamp;
	}

	static of(props: { price: Money; timestamp: Date }): PriceHistoryEntry {
		return new PriceHistoryEntry({
			price: props.price,
			timestamp: props.timestamp,
		});
	}
}
